### DISCUSSION TITLE
Detecting custom and pre-trained classes

### DISCUSSION THREAD

**Discussion by Irikos** [ultralytics/yolov5#4433] [Q&A] [answered]
Hello,

Thanks to the plethora of information available, I've managed to train yolov5 on a custom dataset with significant success, although I just got started with it. However, I want to also detect other classes (such as Person) on the images that I provide, alongside the newly learnt class. Is it possible to do so?

If not, what is the best approach that you recommend? Running 2 detection models (custom and pre-trained) and concatenating the results? Or trying to find a good dataset for the other classes and re-train the custom models for those, as well?

Thanks!

**Comment by glenn-jocher** [comment, marked as answer]
@Irikos your dataset must contain all classes you want to detect if you want to use a single model. You can train on multiple datasets very simply, though they must share the same exact classes, so you could train on COCO (class 0-79) and then relabel your data from class 80 onward, and then train on both, i.e.:
https://github.com/ultralytics/yolov5/blob/dbc06ce29298de4a55d73fa37362dd51a03035ed/data/GlobalWheat2020.yaml#L10-L27

**Comment by Irikos** [comment]
Hey Glenn, thanks for the answer! It helped me better understand how it works. My reply comes so late as I toyed with all the models with different settings. I tried a bit training a model from scratch on person class / images and my own images, but with poor results. I did this by removing all images that do not contain person and removing all other labels from the images that do contain person. To these, I added the my custom dataset the same folders, respecting naming and class ordering.

I assume the results are not that good in detecting my custom dataset because I only have aprox 3000 images of my custom class while the person class is present in well over 60k images. Am I correct in assuming that?

I'm fine with using separate models for detection and just concatenating the results, though.

Thanks!

**Comment by glenn-jocher** [comment]
@Irikos you may want to try to reduce your oversampled classes, or simply train longer to increase exposure to underrepresented classes, as class imbalance issues seem to improve over time when training.

**Comment by laminggg** [comment]
@glenn-jocher hello, I'm wondering if I train a model with 2 classes person and the other class based on yolov5s6.pt, and I have only about 2000+ images of person class, these images comes from 3 people and all shoted indoors, can i get batter performance of detecting person? or get worse?

**Reply by glenn-jocher** [reply]
@laminggg larger datasets will almost always perform better than smaller datasets, though if your class indices align you can also combine multiple datasets during training.

**Reply by laminggg** [reply]
thanks for your reply.

**Comment by sam1006** [comment]
@Irikos Hi Irikos, do you have a trained model for YoloV5 using COCO-person class only? If yes, could you please share? I am trying to detect only the "person" class using YoloV5. The default trained model (trained for 80 classes) is not performing well. So, I assume a trained model for person class only will improve the accuracy. 
Thanks.

**Reply by glenn-jocher** [reply]
@sam1006 you can filter detections by classes easily, so if you just need a person detector you can use the official pretrained models, i.e.:

```bash
python detect.py --classes 0
```

with PyTorch Hub models you can update the classes attribute to get the same effect, i.e.:

```python
model.conf = 0.25  # NMS confidence threshold
      iou = 0.45  # NMS IoU threshold
      agnostic = False  # NMS class-agnostic
      multi_label = False  # NMS multiple labels per box
      classes = None  # (optional list) filter by class, i.e. = [0, 15, 16] for COCO persons, cats and dogs
      max_det = 1000  # maximum number of detections per image
      amp = False  # Automatic Mixed Precision (AMP) inference

results = model(im, size=320)  # custom inference size
```

**Reply by sam1006** [reply]
@glenn-jocher Thank you so much for swift responses. I am currently using "--classes 0" for "person" detection. It works fine, however, I am getting a lot of false positives. Do we have any way/option to reduce false positives?  please guide.
Also do you think if I extract only person related data from COCO dataset and include maybe few hundred images (and labels) from my data and then train the model will improve the results?

**Reply by glenn-jocher** [reply]
@sam1006 yes you might want to train on a person-only dataset, or you can also train COCO only on the persons class, in which case it will perform better at detecting persons than the default COCO dataset.

**Reply by sam1006** [reply]
Thank you so much for guiding!

**Reply by Irikos** [reply]
hey @sam1006 , sorry, I do not. I had a model trained on my images + person images, but deleted it in the meantime. I might train it in the near future though. If you'll still need it then, I'll gladly send it to you.
